### PR TITLE
[SPARK-48665][PYTHON][CONNECT] Support providing a dict in pyspark lit to create a map.

### PR DESCRIPTION
**Pull request by Ronserruya**
### What changes were proposed in this pull request?
Added the option to provide a dict to `pyspark.sql.functions.lit` in order to create a map

### Why are the changes needed?
To make it easier to create a map in pyspark.
Currently, it is only possible via `create_map` which requires a sequence of key,value,key,value... 
Scala already supports such functionality using `typedLit`

A [similar PR](https://github.com/apache/spark/pull/37722) was done in the past to add similar functionality for the creating of an array using a list, so I tried to follow all the changes done there as well.

### Does this PR introduce _any_ user-facing change?
Yes, docstring of `lit` was edited, and new functionality was added

Before:
```python
from pyspark.sql import functions as F
F.lit({"a":1})
# pyspark.errors.exceptions.captured.SparkRuntimeException: [UNSUPPORTED_FEATURE.LITERAL_TYPE] The feature is not supported: Literal for '{asd=2}' of class java.util.HashMap.
```
After:
```python
from pyspark.sql import functions as F
F.lit({"a":1, "b": 2})
# Column<'map(a, 1, b, 2)'>
```

### How was this patch tested?
Manual tests + unittest in CI

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Comment by Ronserruya** [comment]
CC @itholic @HyukjinKwon @zhengruifeng
tagging you since you authored reviewed the [similar PR](https://github.com/apache/spark/pull/37722) for the support of list :)

File: python/pyspark/sql/connect/functions/builtin.py [hunk 1]
```diff
@@ -265,6 +266,8 @@ def lit(col: Any) -> Column:
                 error_class="COLUMN_IN_LIST", message_parameters={"func_name": "lit"}
             )
         return array(*[lit(c) for c in col])
+    elif isinstance(col, dict):
+        return create_map(*[lit(x) for x in chain(*col.items())])
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/functions/builtin.py:270]: the problem is that if we want this as a struct or map

Comment (HyukjinKwon) [reply python/pyspark/sql/connect/functions/builtin.py:270]: Can you compare with createDataFrame? We also have a configuration `spark.sql.pyspark.inferNestedDictAsStruct.enabled`. We might need a configuration to control it.

Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions/builtin.py:270]: it seems this is not supported in PySpark Classic.

Comment (zhengruifeng) [reply python/pyspark/sql/connect/functions/builtin.py:270]: another problem is the result of `create_map` is a `CreateMap` expression, not a literal

Comment (Ronserruya) [reply python/pyspark/sql/connect/functions/builtin.py:270]: > it seems this is not supported in PySpark Classic.

Sorry, I'm not sure what you meant here, pyspark "Classic" means local pyspark (as opposed to connect)?
I added the code change to both the classic functions and the connect ones.

> another problem is the result of `create_map` is a `CreateMap` expression, not a literal

Is that a problem? the `array` function that is used above in the same function to create an array "literal" also returns a `CreateArray` expression and not an array literal

Comment (allisonwang-db) [reply python/pyspark/sql/connect/functions/builtin.py:270]: Should we also check if any of the dictionary values contain a Column type object and throw similar exceptions as in the array case above?

Comment (Ronserruya) [reply python/pyspark/sql/connect/functions/builtin.py:270]: > Should we also check if any of the dictionary values contain a Column type object and throw similar exceptions as in the array case above?

Sure, I'll add that check + the appropriate error

Comment (Shiva0085) [reply python/pyspark/sql/connect/functions/builtin.py:276]: So, I think we can merge the changes with master.

**Pull request closed without merging**